proposals/NNNN-importing-forward-declared-objc-interfaces-and-protocols.md

```diff
@@ -0,0 +1,177 @@
+# Importing Forward Declared Objective-C Interfaces and Protocols
+
+* Proposal: [SE-NNNN](NNNN-importing-forward-declared-objc-interfaces-and-protocols.md)
+* Authors: [Nuri Amari](https://github.com/NuriAmari)
+* Review Manager: TBD
+* Implementation: https://github.com/apple/swift/pull/61606
+
+## Introduction
+
+This proposal seeks to improve the usability of existing Objective-C libraries from Swift by reducing the negative
+impact forward declarations have on API visibility from Swift. We wish to start synthesizing placeholder types to
+represent forward declared Objective-C interfaces and protocols in Swift.
+
+Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/t/pitch-importing-forward-declared-objective-c-classes-and-protocols/61926)
+
+## Motivation
+
+Forward declarations are very common in many existing Objective-C code bases, used often to break cyclic dependencies or to improve build performance.
+Unfortunately, when it comes to "importability" into Swift they are quite detrimental.
+
+As it stands, the ClangImporter will fail to import any declaration that references a forward declared type in many common cases. This means a single
+forward declared type can render larger portions of an Objective-C API unusable from Swift. For example, the following Objective-C API from the implementation PR
+is empty from the Swift perspective (Swift textual interface generated via swift-ide-test):
+
+_Objective-C_
+```objective-c
+#import <Foundation/Foundation.h>
+
+@class ForwardDeclaredInterface;
+@protocol ForwardDeclaredProtocol;
+
+@interface IncompleteTypeConsumer1 : NSObject
+@property id<ForwardDeclaredProtocol> propertyUsingAForwardDeclaredProtocol1;
+@property ForwardDeclaredInterface *propertyUsingAForwardDeclaredInterface1;
+- (id)init;
+- (NSObject<ForwardDeclaredProtocol> *)methodReturningForwardDeclaredProtocol1;
+- (ForwardDeclaredInterface *)methodReturningForwardDeclaredInterface1;
+- (void)methodTakingAForwardDeclaredProtocol1:
+    (id<ForwardDeclaredProtocol>)param;
+- (void)methodTakingAForwardDeclaredInterface1:
+            (ForwardDeclaredInterface *)param;
+@end
+
+ForwardDeclaredInterface *CFunctionReturningAForwardDeclaredInterface1();
+void CFunctionTakingAForwardDeclaredInterface1(
+    ForwardDeclaredInterface *param);
+
+NSObject<ForwardDeclaredProtocol> *CFunctionReturningAForwardDeclaredProtocol1();
+void CFunctionTakingAForwardDeclaredProtocol1(
+    id<ForwardDeclaredProtocol> param);
+```
+
+_Swift_
+```swift
+class IncompleteTypeConsumer1 : NSObject {
+  init!()
+}
+```
+We wish to make the experience of consuming an Objective-C API with forward declarations from Swift more consistent with the experience of consuming the API from Objective-C.
```

Comment:
You should call out up-front that the user _can_ resolve this today by importing the module providing the full definition of the type, but that the compiler's current behavior doesn't give any feedback that this is what they should do. Emphasize the point here explicitly that this proposal gives the user that important feedback that is currently missing.

Reply:
Relating to this comment, and the one above about silently dropping declarations: That isn't strictly true, but moreover in my eyes, increased feedback isn't really the goal of this proposal either.

A year or so ago I added diagnostics to the ClangImporter to diagnose these kinds of declarations being dropped. It has been enabled by default for some time: https://forums.swift.org/t/pitch-lazy-clangimporter-diagnostics-enabled-by-default/54651. Notably, these diagnostics are lazy meaning they are only reported if you attempt to use an "unimportable" entity. In that case, you'll get an error like this:

```
/home/nuria/Git/swift-project/swift/scratch/test.h:55:1: warning: method 'methodReturningForwardDeclaredInterface' not imported
- (ForwardDeclaredInterface *) methodReturningForwardDeclaredInterface;
^
/home/nuria/Git/swift-project/swift/scratch/test.h:55:1: note: return type not imported
- (ForwardDeclaredInterface *) methodReturningForwardDeclaredInterface;
^
/home/nuria/Git/swift-project/swift/scratch/test.h:55:1: note: interface 'ForwardDeclaredInterface' is incomplete
- (ForwardDeclaredInterface *) methodReturningForwardDeclaredInterface;
^
/home/nuria/Git/swift-project/swift/scratch/test.h:45:1: note: interface 'ForwardDeclaredInterface' forward declared here
@class ForwardDeclaredInterface;
^
```

If you want eager diagnostics (ie. the diagnostic is not predicated on a usage), I also added that, but it is behind a flag, something like `-enable-eager-clang-importer-diagnostics`, I can't recall the exact spelling. 

If you are suggesting that these diagnostics need to be more explicit about the solution, that's doable, but diagnostics was not really the goal for this proposal. Rather the goal here is to actually import the declaration rather than diagnose the failure.

Reply:
Ok, that's reasonable. Thanks for the refresher, I wasn't sure what the current state of things was for your previous work. That being the case, some other readers may not be familiar with the history here either, so I think it would be beneficial to link to your previous work and have a brief statement about the diagnostics already shown and how this proposal builds on that by actually synthesizing the type.